File: src/connectors/wayback.js
import { BaseConnector } from './base.js';
import https from 'https';

export class WaybackConnector extends BaseConnector {
    constructor(config, state, telemetry) {
        super(config, state, telemetry);
        this.name = 'Wayback';
    }

    async run(investigationId, phaseId, inputType, inputValue) {
        const runId = this.state.recordConnectorRun(investigationId, phaseId, this.name, inputType, inputValue);
        this.telemetry.connectorStart(investigationId, this.name, phaseId, { type: inputType, value: inputValue });

        try {
            const entities = [];
            const domain = inputValue.replace(/^https?:\/\//, '');

            const urls = await this._fetch(`https://web.archive.org/cdx/search/cdx?url=${encodeURIComponent(domain)}/*&output=json&fl=original,timestamp,mimetype,statuscode&collapse=urlkey&limit=200`);
            const records = JSON.parse(urls);

            if (Array.isArray(records) && records.length > 1) {
                const seen = new Set();
                for (const row of records.slice(1)) {
                    const [original, timestamp, mime, status] = row;
                    if (seen.has(original)) continue;
                    seen.add(original);

                    entities.push({
                        type: 'url',
                        data: {
                            url: original,
                            timestamp,
                            mime_type: mime,
                            status_code: parseInt(status) || null,
                            source: 'wayback_machine',
                            archive_url: `https://web.archive.org/web/${timestamp}/${original}`
                        },
                        confidence: 0.8
                    });

                    try {
                        const u = new URL(original);
                        const sub = u.hostname;
                        if (sub !== domain && sub.endsWith(domain)) {
                            entities.push({
                                type: 'subdomain',
                                data: { name: sub, parent_domain: domain, source: 'wayback_machine' },
                                confidence: 0.7,
                                asKnown: { type: 'domain', value: sub }
                            });
                        }

                        if (u.pathname.match(/\.(js|json|xml|env|config|bak|sql|log|txt|csv|zip|tar|gz)$/i)) {
                            entities.push({
                                type: 'url',
                                data: {
                                    url: original,
                                    interesting_file: true,
                                    extension: u.pathname.split('.').pop(),
                                    source: 'wayback_machine'
                                },
                                confidence: 0.6
                            });
                        }

                        if (u.pathname.match(/\/api\/|\/v[12]\/|\/graphql|\/rest\//i)) {
                            entities.push({
                                type: 'url',
                                data: { url: original, api_endpoint: true, source: 'wayback_machine' },
                                confidence: 0.7
                            });
                        }
                    } catch {}
                }

                const emailRegex = /[\w.+-]+@[\w-]+\.[\w.-]+/g;
                const urlText = records.slice(1).map(r => r[0]).join('\n');
                for (const match of urlText.matchAll(emailRegex)) {
                    entities.push({
                        type: 'email',
                        data: { address: match[0], source: 'wayback_url_params' },
                        confidence: 0.4,
                        asKnown: { type: 'email', value: match[0] }
                    });
                }
            }

            let newCount = 0;
            for (const entity of entities) {
                const added = this.state.addEntity(investigationId, entity.type, entity.data, entity.confidence || 0.5);
                if (added.new) newCount++;
                this.telemetry.entityDiscovered(investigationId, entity.type, added.new, this.name);
                if (entity.asKnown) {
                    this.state.addKnown(investigationId, entity.asKnown.type, entity.asKnown.value, this.name, entity.confidence || 0.5);
                }
            }

            this.state.completeConnectorRun(runId, 'completed', newCount, null, null, 0);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'completed', entitiesFound: newCount, input: { type: inputType, value: inputValue }
            });
            return { status: 'completed', entities, newCount };
        } catch (e) {
            this.state.completeConnectorRun(runId, 'failed', 0, null, e.message, 1);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'failed', entitiesFound: 0, input: { type: inputType, value: inputValue }
            });
            return { status: 'failed', error: e.message };
        }
    }

    _fetch(url) {
        return new Promise((resolve, reject) => {
            https.get(url, { headers: { 'User-Agent': 'cyclops-osint' }, timeout: 30000 }, res => {
                let data = '';
                res.on('data', chunk => { data += chunk; });
                res.on('end', () => resolve(data));
            }).on('error', reject);
        });
    }

    async healthCheck() { return { ok: true, version: 'native' }; }
}

File: cyclops/connectors/wayback.py
import aiohttp
from cyclops.connectors.base import BaseConnector


class WaybackConnector(BaseConnector):
    def __init__(self, config, state, telemetry):
        super().__init__(config, state, telemetry)
        self.name = "Wayback"

    async def run(self, investigation_id, phase_id, input_type, input_value):
        run_id = self.state.record_connector_run(investigation_id, phase_id, self.name, input_type, input_value)
        self.telemetry.connector_start(investigation_id, self.name, phase_id, {"type": input_type, "value": input_value})
        try:
            entities = []
            url = f"https://web.archive.org/cdx/search/cdx?url=*.{input_value}&output=json&fl=original&collapse=urlkey&limit=100"
            async with aiohttp.ClientSession() as session:
                async with session.get(url, timeout=aiohttp.ClientTimeout(total=30)) as resp:
                    if resp.status >= 400:
                        raise Exception(f"Wayback HTTP {resp.status}")
                    data = await resp.json(content_type=None)
            seen = set()
            for row in (data or [])[1:]:
                u = row[0] if row else ""
                if u and u not in seen:
                    seen.add(u)
                    entities.append({"type": "url", "data": {"url": u, "source": "wayback"}, "confidence": 0.6})
            new_count = 0
            for entity in entities:
                added = self.state.add_entity(investigation_id, entity["type"], entity["data"], entity.get("confidence", 0.5))
                if added["new"]:
                    new_count += 1
                self.telemetry.entity_discovered(investigation_id, entity["type"], added["new"], self.name)
            self.state.complete_connector_run(run_id, "completed", new_count)
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "completed", "entitiesFound": new_count, "input": {"type": input_type, "value": input_value}})
            return {"status": "completed", "entities": entities, "newCount": new_count}
        except Exception as e:
            self.state.complete_connector_run(run_id, "failed", 0, error=str(e))
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "failed", "input": {"type": input_type, "value": input_value}})
            return {"status": "failed", "error": str(e)}

    async def health_check(self):
        return {"ok": True, "version": "cdx"}
